File: Serene/Serene.Script/Administration/User/UserDialog.cs
using Serenity;
using System.Runtime.CompilerServices;

namespace Serene.Administration
{
    [Imported]
    public class UserDialog : EntityDialog<UserRow>
    {
    }

    /* 
    This class has been ported to TypeScript. See UserDialog.ts
    Code below is only a reference for those who want to use Saltaralle

    [IdProperty("UserId"), NameProperty("Username"), IsActiveProperty("IsActive")]
    [FormKey("Administration.User"), LocalTextPrefix("Administration.User"), Service("Administration/User")]
    public class UserDialog : EntityDialog<UserRow>
    {
        private UserForm form;

        public UserDialog()
        {
            form = new UserForm(this.IdPrefix);

            form.Password.AddValidationRule(this.uniqueName, e =>
            {
                if (form.Password.Value.Length < 7)
                    return "Password must be at least 7 characters!";

                return null;
            });

            form.PasswordConfirm.AddValidationRule(this.uniqueName, e =>
            {
                if (form.Password.Value != form.PasswordConfirm.Value)
                    return "The passwords entered doesn't match!";

                return null;
            });
        }

        protected override List<ToolButton> GetToolbarButtons()
        {
            var buttons = base.GetToolbarButtons();

            buttons.Add(new ToolButton
            {
                Title = Q.Text("Site.UserDialog.EditRolesButton"),
                CssClass = "users-button",
                OnClick = delegate
                {
                    new UserRoleDialog(new UserRoleDialogOptions
                    {
                        UserID = this.Entity.UserId.Value,
                        Username = this.Entity.Username,
                    }).DialogOpen();
                }
            });

            buttons.Add(new ToolButton
            {
                Title = Q.Text("Site.UserDialog.EditPermissionsButton"),
                CssClass = "lock-button",
                OnClick = delegate
                {
                    new UserPermissionDialog(new UserPermissionDialogOptions
                    {
                        UserID = this.Entity.UserId.Value,
                        Username = this.Entity.Username,
                    }).DialogOpen();
                }
            });

            return buttons;
        }

        protected override void UpdateInterface()
        {
            base.UpdateInterface();

            toolbar.FindButton("users-button").ToggleClass("disabled", this.IsNewOrDeleted);
            toolbar.FindButton("lock-button").ToggleClass("disabled", this.IsNewOrDeleted);
        }
    }
    */
}

File: Serene/Serene.Web/Modules/Administration/User/UserDialog.ts
namespace Serene.Administration {

    @Serenity.Decorators.registerClass()
    export class UserDialog extends Serenity.EntityDialog<UserRow, any> {
        protected getFormKey() { return UserForm.formKey; }
        protected getIdProperty() { return UserRow.idProperty; }
        protected getIsActiveProperty() { return UserRow.isActiveProperty; }
        protected getLocalTextPrefix() { return UserRow.localTextPrefix; }
        protected getNameProperty() { return UserRow.nameProperty; }
        protected getService() { return UserService.baseUrl; }

        protected form = new UserForm(this.idPrefix);

        constructor() {
            super();

            this.form.Password().addValidationRule(this.uniqueName, e => {
                if (this.form.Password().get_value().length < 7)
                    return "Password must be at least 7 characters!";
            });

            this.form.PasswordConfirm().addValidationRule(this.uniqueName, e => {
                if (this.form.Password().get_value() != this.form.PasswordConfirm().get_value())
                    return "The passwords entered doesn't match!";
            });
        }

        protected getToolbarButtons()
        {
            let buttons = super.getToolbarButtons();

            buttons.push({
                title: Q.text('Site.UserDialog.EditRolesButton'),
                cssClass: 'users-button',
                onClick: () =>
                {
                    new UserRoleDialog({
                        userID: this.entity.UserId,
                        username: this.entity.Username
                    }).dialogOpen();
                }
            });

            buttons.push({
                title: Q.text('Site.UserDialog.EditPermissionsButton'),
                cssClass: 'lock-button',
                onClick: () =>
                {
                    new UserPermissionDialog({
                        userID: this.entity.UserId,
                        username: this.entity.Username
                    }).dialogOpen();
                }
            });

            return buttons;
        }

        protected updateInterface() {
            super.updateInterface();

            this.toolbar.findButton('users-button').toggleClass('disabled', this.isNewOrDeleted());
            this.toolbar.findButton("lock-button").toggleClass("disabled", this.isNewOrDeleted());
        }

        protected afterLoadEntity() {
            super.afterLoadEntity();

            // these fields are only required in new record mode
            this.form.Password().element.toggleClass('required', this.isNew())
                .closest('.field').find('sup').toggle(this.isNew());
            this.form.PasswordConfirm().element.toggleClass('required', this.isNew())
                .closest('.field').find('sup').toggle(this.isNew());
        }
    }
}